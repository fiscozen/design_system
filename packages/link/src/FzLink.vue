<template>
  <span v-if="disabled" :class="spanClass">
    <slot></slot>
  </span>
  <a v-else-if="external" :href="to.toString()" :class="linkClass" :target>
    <slot></slot>
  </a>
  <router-link v-else :to :replace :class="linkClass" :target>
    <slot></slot>
  </router-link>
</template>

<script setup lang="ts">
import { computed } from 'vue'
import { FzLinkProps } from './types'

const props = withDefaults(defineProps<FzLinkProps>(), {
  type: 'default',
  style: 'default',
  size: 'lg',
  disabled: false,
  replace: false,
  external: false,
})

const commonClass = computed(() => [
  'border-1 border-transparent',
  {
    'text-xs': props.size === 'xs',
    'text-sm': props.size === 'sm',
    'text-md': props.size === 'md',
    'text-lg': props.size === 'lg',
    underline: props.style === 'underline'
  }
])

const linkClass = computed(() => [
  ...commonClass.value,
  'hover:underline',
  {
    'text-blue-500 hover:text-blue-600 focus:text-blue-600 focus:border-blue-600':
      props.type === 'default',
    'text-semantic-error hover:text-red-600 focus:text-red-600 focus:border-red-600':
      props.type === 'danger'
  }
])

const spanClass = computed(() => [
  ...commonClass.value,
  'cursor-not-allowed',
  {
    'text-red-200': props.type === 'danger',
    'text-blue-200': props.type === 'default'
  }
])
</script>
