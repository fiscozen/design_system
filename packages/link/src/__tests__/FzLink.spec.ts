import { describe, it } from 'vitest'
import { mount } from '@vue/test-utils'
import FzLink from '../FzLink.vue'
import { createRouter, createWebHistory } from 'vue-router'

const router = createRouter({
  history: createWebHistory(),
  routes: [{ name: '', path: '/example', component: () => {} }]
})

describe.concurrent('FzLink', () => {
  it('image matches snapshot', async ({ expect }) => {
    const wrapper = mount(FzLink, {
      props: {
        to: '/example'
      },
      slots: {
        default: 'This is a link'
      },
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render link when not disabled', async ({ expect }) => {
    const wrapper = mount(FzLink, {
      props: {
        to: '/example'
      },
      slots: {
        default: 'This is a link'
      },
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.find('a').exists()).toBe(true)
  })

  it('should render span when disabled', async ({ expect }) => {
    const wrapper = mount(FzLink, {
      props: {
        to: '/example',
        disabled: true
      },
      slots: {
        default: 'This is a link'
      },
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.find('span').exists()).toBe(true)
  })

  it('should render link href', async ({ expect }) => {
    const wrapper = mount(FzLink, {
      props: {
        to: '/example'
      },
      slots: {
        default: 'This is a link'
      },
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.find('a').attributes('href')).toBe('/example')
  })

  it('should render text content', async ({ expect }) => {
    const wrapper = mount(FzLink, {
      props: {
        to: '/example'
      },
      slots: {
        default: 'This is a link'
      },
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.html()).toContain('This is a link')
  })

  it('should render size', async ({ expect }) => {
    const wrapper = mount(FzLink, {
      props: {
        size: 'md',
        to: '/example'
      },
      slots: {
        default: 'This is a link'
      },
      global: {
        plugins: [router]
      }
    })

    expect(wrapper.find('a').classes()).toContain('text-md')
  })
})
