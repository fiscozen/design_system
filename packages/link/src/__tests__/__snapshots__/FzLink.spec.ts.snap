// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FzLink > image matches snapshot 1`] = `"<a href="/example" class="border-1 border-transparent inline-block text-lg hover:underline text-blue-500 hover:text-blue-600 focus:text-blue-600 focus:border-blue-600">This is a link</a>"`;
